File: java/scf-bp-component/src/main/java/org/mydotey/scf/bp/component/labeled/LabeledPropertyComponent.java
package org.mydotey.scf.bp.component.labeled;

import java.util.ArrayList;

import org.mydotey.scf.Property;
import org.mydotey.scf.facade.LabeledConfigurationProperties;
import org.mydotey.scf.facade.LabeledStringProperties;
import org.mydotey.scf.labeled.LabeledConfigurationManager;
import org.mydotey.scf.labeled.LabeledKey;
import org.mydotey.scf.labeled.PropertyLabel;
import org.mydotey.scf.labeled.PropertyLabels;

/**
 * @author koqizhao
 *
 * Jul 25, 2018
 */
public class LabeledPropertyComponent {

    private PropertyLabels _propertyLabels;

    private LabeledStringProperties _properties;

    private Property<LabeledKey<String>, Integer> _requestTimeout;

    public LabeledPropertyComponent(LabeledConfigurationManager manager) {
        _properties = new LabeledStringProperties(manager);
        initLabels();

        _requestTimeout = _properties.getIntProperty(toLabeledKey("request.timeout"), 1000);
    }

    // priority: (key, app, dc) > (key, app) > (key, dc) > key
    private void initLabels() {
        PropertyLabel dcLabel = LabeledConfigurationProperties.newLabel("dc", "shanghai");
        PropertyLabel appLabel = LabeledConfigurationProperties.newLabel("app", "10000");
        ArrayList<PropertyLabel> dcLabels = new ArrayList<>();
        dcLabels.add(dcLabel);
        ArrayList<PropertyLabel> appLabels = new ArrayList<>();
        appLabels.add(appLabel);
        ArrayList<PropertyLabel> allLabels = new ArrayList<>();
        allLabels.add(dcLabel);
        allLabels.add(appLabel);

        PropertyLabels alternative1 = LabeledConfigurationProperties.newLabels(dcLabels, PropertyLabels.EMPTY);
        PropertyLabels alternative2 = LabeledConfigurationProperties.newLabels(appLabels, alternative1);
        _propertyLabels = LabeledConfigurationProperties.newLabels(allLabels, alternative2);
    }

    protected LabeledKey<String> toLabeledKey(String key) {
        return LabeledConfigurationProperties.<String> newKeyBuilder().setKey(key).setPropertyLabels(_propertyLabels)
                .build();
    }

    /**
     * use the following methods to get your properties / property values
     *      LabeledKey&lt;String&gt; labeledKey = toLabeledKey(key);
     *      getProperties().getSomeProperty
     *      getProperties().getSomePropertyValue
     *      getProperties().getManager()
     */
    protected LabeledStringProperties getProperties() {
        return _properties;
    }

    public void yourComponentApi() {
        System.out.println("OuterManagerComponent is doing something");
        System.out.println("request.timeout is: " + _requestTimeout.getValue());
    }

}

File: dotnet/src/MyDotey.SCF.BP.Component/Labeled/LabeledPropertyComponent.cs
using System;
using System.Collections.Generic;

using MyDotey.SCF.Facade;
using MyDotey.SCF.Labeled;

namespace MyDotey.SCF.BP.Component.Labeled
{
    /**
     * @author koqizhao
     *
     * Jul 25, 2018
     */
    public class LabeledPropertyComponent
    {
        private PropertyLabels _propertyLabels;

        private LabeledStringProperties _properties;

        private IProperty<LabeledKey<string>, int?> _requestTimeout;

        public LabeledPropertyComponent(ILabeledConfigurationManager manager)
        {
            _properties = new LabeledStringProperties(manager);
            InitLabels();

            _requestTimeout = _properties.GetIntProperty(ToLabeledKey("request.timeout"), 1000);
        }

        // priority: (key, app, dc) > (key, app) > (key, dc) > key
        private void InitLabels()
        {
            IPropertyLabel dcLabel = LabeledConfigurationProperties.NewLabel("dc", "shanghai");
            IPropertyLabel appLabel = LabeledConfigurationProperties.NewLabel("app", "10000");
            List<IPropertyLabel> dcLabels = new List<IPropertyLabel>();
            dcLabels.Add(dcLabel);
            List<IPropertyLabel> appLabels = new List<IPropertyLabel>();
            appLabels.Add(appLabel);
            List<IPropertyLabel> allLabels = new List<IPropertyLabel>();
            allLabels.Add(dcLabel);
            allLabels.Add(appLabel);

            PropertyLabels alternative1 = LabeledConfigurationProperties.NewLabels(dcLabels, PropertyLabels.EMPTY);
            PropertyLabels alternative2 = LabeledConfigurationProperties.NewLabels(appLabels, alternative1);
            _propertyLabels = LabeledConfigurationProperties.NewLabels(allLabels, alternative2);
        }

        protected LabeledKey<string> ToLabeledKey(string key)
        {
            return LabeledConfigurationProperties.NewKeyBuilder<string>().SetKey(key).SetPropertyLabels(_propertyLabels)
                    .Build();
        }

        /**
         * use the following methods to get your properties / property values
         *      LabeledKey&lt;string&gt; labeledKey = ToLabeledKey(key);
         *      Properties.GetSomeProperty
         *      Properties.GetSomePropertyValue
         *      Properties.Manager
         */
        protected LabeledStringProperties Properties { get { return _properties; } }

        public void YourComponentApi()
        {
            Console.WriteLine("OuterManagerComponent is doing something");
            Console.WriteLine("request.timeout is: " + _requestTimeout.Value);
        }
    }
}